nav2_waypoint_follower/include/nav2_waypoint_follower/waypoint_follower.hpp

```diff
@@ -115,16 +126,62 @@ class WaypointFollower : public nav2_util::LifecycleNode
    */
   void goalResponseCallback(const rclcpp_action::ClientGoalHandle<ClientT>::SharedPtr & goal);
 
-  // Our action server
+  /**
+   * @brief send robot through each of GPS
+   *        point , which are converted to map frame first then using a client to
+   *        `FollowWaypoints` action.
+   *
+   * @param waypoints, acquired from action client
+   */
+  void followGPSWaypoints();
+
+/**
+ * @brief given some gps_waypoints, converts them to map frame using robot_localization's service `fromLL`.
+ *        Constructs a vector of stamped poses in map frame and returns them.
+ *
+ * @param gps_waypoints
+ * @param parent_node
+ * @param fromll_client
+ * @return std::vector<geometry_msgs::msg::PoseStamped>
+ */
+  static std::vector<geometry_msgs::msg::PoseStamped> convertGPSWaypointstoPosesinMap(
+    const
+    std::vector<sensor_msgs::msg::NavSatFix> & gps_waypoints,
+    const rclcpp_lifecycle::LifecycleNode::SharedPtr & parent_node,
+    const rclcpp::Client<robot_localization::srv::FromLL>::SharedPtr & fromll_client);
+
+  /**
+   * @brief Action client result callback
+   * @param result Result of action server updated asynchronously
+   */
+  void GPSResultCallback(const rclcpp_action::ClientGoalHandle<ClientTGPS>::WrappedResult & result);
+
+  /**
+   * @brief Action client goal response callback
+   * @param goal Response of action server updated asynchronously
+   */
+  void GPSGoalResponseCallback(const rclcpp_action::ClientGoalHandle<ClientTGPS>::SharedPtr & goal);
+
+  // Our action server for waypoint following
   std::unique_ptr<ActionServer> action_server_;
   ActionClient::SharedPtr nav_to_pose_client_;
-  rclcpp::Node::SharedPtr client_node_;
+  rclcpp::Node::SharedPtr nav_to_pose_client_node_;
   std::shared_future<rclcpp_action::ClientGoalHandle<ClientT>::SharedPtr> future_goal_handle_;
   bool stop_on_failure_;
   ActionStatus current_goal_status_;
   int loop_rate_;
   std::vector<int> failed_ids_;
 
+  // Our action server for GPS waypoint following
+  std::unique_ptr<ActionServerGPS> gps_action_server_;
+  rclcpp::Client<robot_localization::srv::FromLL>::SharedPtr from_ll_to_map_client_;
+  ActionClientGPS::SharedPtr waypoint_follower_action_client_;
+  rclcpp::Node::SharedPtr waypoint_follower_client_node_;
+  ActionStatus gps_current_goal_status_;
+  ClientTGPS::Goal waypoint_follower_goal_;
+  WaypointFollowerGoalHandle::SharedPtr waypoint_follower_goal_handle_;
+  rclcpp::CallbackGroup::SharedPtr callback_group_;
```

Comment:
Why a unique callback group? It only has a single thing in it, so its already in a callback group when you create something.

Reply:
I previously didn't use callback group, but since this service was being called inside another action callback(`FollowGPSWaypoints`), I was running into an issue indicating that I am spinning the node twice, which was true, because I was doing `spin_until_future_complete()` to wait result of service call as well as action call. `MutuallyExclusiveCallbackGroup` allowed one of the callbacks to be executed at one time so it addressed the issue. 
I see that `ServiceCallback` also internally does `spin_until_future_complete()`  but I couldn't understand why it does not trigger that error, since it is the same thing. But I am happy that it is performing ok with `nav2_utils::ServiceClient`